### PR TITLE
Map transports not services

### PR DESCRIPTION
**Pull request by rklaehn**
TLDR: instead of mapping the service, map the transport. This gets rid of the mapper in the service and makes the RpcClient/RpcChannel and friends again just pure newtype like things.

To make this work I had to make a bigger change. The various channel traits now have associated types `In` and `Out`, instead of type parameters.

- The three service types (RpcClient, RpcServer and RpcChannel) are now again pure newtypes over the underlying transports that just add the service type.

- map is available on RpcClient and RpcChannel, but unlike before it maps the underlying transport, so there is no third type parameter `SC` needed anymore

- the boxed transport is available by default, .boxed() is available on all transports including those that are already boxed. For RpcClient, RpcChannel and RpcServer, the transport type `C` is set by default to the boxed transport. That way in many cases you can avoid having to provide the transport type parameter at all, at the expense of a tiny bit of performance.

- also made the combined transport available by default - it does not have any dependencies, so it seems like having a feature flag for it is overkill.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -53,23 +53,15 @@ impl<SC, S, C: Clone> Clone for RpcClient<S, C, SC> {
 /// that support it, [crate::message::ClientStreaming] and [crate::message::BidiStreaming].
 #[pin_project]
 #[derive(Debug)]
-pub struct UpdateSink<SC, C, T, S = SC>(
```
Comment (rklaehn) [inline src/client.rs:56]: Weird that the phantomdata is pub -> fixed

**Comment by dignifiedquire** [comment]
there is a lot of changes, is there a specific file showing the benefits of this change, to understand how it helps the user?

**Comment by rklaehn** [comment]
> there is a lot of changes, is there a specific file showing the benefits of this change, to understand how it helps the user?

The closest thing to that is the smoke "test" in mapped.rs

As for a real use case: when extracting the rpc for blobs, I had the following problem. Either you have to drag around the service type parameter matching the transport channel `SC` everywhere, or you have to define a type alias. With this PR you simply do not have the `SC` type parameter anymore, so no need to drag it around.

E.g. you got blobs::Client::new. You can either make it take a RpcClient<BlobsService>, having no type parameters at all. Then you create one by saying blobs::Client::new(iroh_rpc_client.map().boxed()).

...or take a RpcClient<BlobsService, C> where C is a channel that is compatible with BlobsService and defaults to BoxedServiceConnection<BlobsService>. Then you can call it with blobs::Client::new(iroh_rpc_client.map()) and don't have the (small) boxing overhead.

In any case, you avoid having to have an open S: Service parameter at all in the iroh blobs client and service implementation, which at least while writing it I found *horribly* confusing...

File: examples/modularize.rs [hunk 2]
```diff
@@ -19,14 +21,14 @@ use app::AppService;
...
-    let (server_conn, client_conn) = flume::service_connection::<AppService>(1);
+    let (server_conn, client_conn) = flume::connection(1);
 
     // spawn the server
     let handler = app::Handler::default();
     tokio::task::spawn(run_server(server_conn, handler));
 
     // run a client demo
-    client_demo(client_conn).await?;
+    client_demo(BoxedServiceConnection::<AppService>::new(client_conn)).await?;
```
Comment (dignifiedquire) [inline examples/modularize.rs:31]: what's the difference between a `connection` and a `service connection`?
